buildSrc/src/main/kotlin/FetchLatestGradleVersionTask.kt

```diff
@@ -0,0 +1,104 @@
+abstract class FetchLatestGradleVersionTask : DefaultTask() {
+
+    @get:OutputFile
+    abstract val versionFile: RegularFileProperty
+
+    init {
+        group = "gradle wrapper"
+        description = "Fetches the latest Gradle version (comparing stable release and RC)"
+    }
+
+    @TaskAction
+    fun fetchVersion() {
+        val currentRelease = fetchVersionFromEndpoint("$HTTPS_SERVICES_GRADLE_ORG_VERSIONS/current")
+        val releaseCandidate = fetchVersionFromEndpoint("$HTTPS_SERVICES_GRADLE_ORG_VERSIONS/release-candidate")
+
+        logger.lifecycle("Latest stable release: $currentRelease")
+        logger.lifecycle("Latest release candidate: ${releaseCandidate ?: "none"}")
+
+        val selectedVersion = when {
+            currentRelease == null -> {
+                throw IllegalStateException("Failed to fetch current Gradle release version")
+            }
+
+            releaseCandidate == null -> {
+                logger.lifecycle("No RC available, using stable release: $currentRelease")
+                currentRelease
+            }
+
+            else -> {
+                val newerVersion = compareVersions(currentRelease, releaseCandidate)
+                logger.lifecycle("Selected newer version: $newerVersion")
+                newerVersion
+            }
+        }
+
+        versionFile.get().asFile.writeText(selectedVersion)
+        logger.lifecycle("Written version to file: $selectedVersion")
+    }
+
+    private fun fetchVersionFromEndpoint(urlString: String): String? {
+        return try {
+            val jsonResponse = URI(urlString).toURL().readText()
+            parseVersion(jsonResponse)
+        } catch (e: java.io.IOException) {
+            throw IllegalStateException(
+                "Failed to fetch Gradle version from $urlString",
+                e
+            )
+        } catch (e: JsonSyntaxException) {
+            throw IllegalStateException(
+                "Failed to parse Gradle version response from $urlString",
+                e
+            )
+        }
+    }
+
+    private fun parseVersion(jsonResponse: String): String? {
+        val gson = Gson()
+        val versionInfo = gson.fromJson(jsonResponse, GradleVersionInfo::class.java)
+        return versionInfo?.version
+    }
+
+    private fun compareVersions(version1: String, version2: String): String {
+        val v1 = VersionNumber.parse(version1)
+        val v2 = VersionNumber.parse(version2)
+
+        return when {
+            v1 > v2 -> {
+                logger.lifecycle("$version1 is newer than $version2")
+                version1
+            }
+
+            v2 > v1 -> {
+                logger.lifecycle("$version2 is newer than $version1")
+                version2
+            }
+
+            else -> {
+                logger.lifecycle("Versions are equal, using first: $version1")
+                version1
+            }
```

Comment:
💅 These three logging statement are useless given the ones above in `fetchVersions()` already give the result. The `when` could be largely reduced by removing them:

```kotlin
return when {
    v1 > v2 -> version1
    v2 > v1 -> version2
    else -> version1
}
```